src/PAModelTests/DataSourceTests.cs

```diff
@@ -117,6 +120,142 @@ public void TestConnectionInstanceIDHandling(string appName)
+            errors = msApp.SaveToSources(sources2.Dir, msAppTemp.FullPath);
+            errors.ThrowOnErrors();
+
+            Assert.IsFalse(msApp._entropy.WasUnusedDataSourcesForLocalDbRefsAbsent());
+            Assert.IsNull(msApp._entropy.GetUnusedDataSourcesForLocalDbRef("default.cds"));
+            Assert.IsTrue(msApp._entropy.UnusedDataSourcesForLocalDbRefs.ContainsKey("default.cds"));
+        }
+
+        [DataTestMethod]
+        [DataRow("UnusedDataSourcesHashMismatch.msapp")]
+        public void TestWhenDataSourcesIsSetEmptyDictionary(string appName)
+        {
+            var pathToMsApp = Path.Combine(Environment.CurrentDirectory, "Apps", appName);
+            Assert.IsTrue(File.Exists(pathToMsApp));
+
+            var (msApp, errors) = CanvasDocument.LoadFromMsapp(pathToMsApp);
+            msApp._dataSourceReferences["default.cds"].dataSources = new Dictionary<string, LocalDatabaseReferenceDataSource>();
+            errors.ThrowOnErrors();
+
+            using var sourcesTempDir = new TempDir();
+            var sourcesTempDirPath = sourcesTempDir.Dir;
+            errors = msApp.SaveToSources(sourcesTempDirPath, pathToMsApp);
+            (msApp, errors) = CanvasDocument.LoadFromSources(sourcesTempDirPath);
+            using var msAppTemp = new TempFile();
+            using var sources2 = new TempDir();
+            var errors2 = new ErrorContainer();
+            MsAppSerializer.SaveAsMsApp(msApp, msAppTemp.FullPath, errors2);
+            errors = msApp.SaveToSources(sources2.Dir, msAppTemp.FullPath);
+            errors.ThrowOnErrors();
+
+            Assert.IsFalse(msApp._entropy.WasUnusedDataSourcesForLocalDbRefsAbsent());
+            Assert.IsTrue(msApp._entropy.GetUnusedDataSourcesForLocalDbRef("default.cds").Count == 0);
+        }
+
+        [DataTestMethod]
+        [DataRow("UnusedDataSourcesHashMismatch.msapp")]
+        public void TestWhenDataSourcesAreNotPresentBackwardCompat(string appName)
+        {
+            var pathToMsApp = Path.Combine(Environment.CurrentDirectory, "Apps", appName);
+            Assert.IsTrue(File.Exists(pathToMsApp));
+
+            var (msApp, errors) = CanvasDocument.LoadFromMsapp(pathToMsApp);
+            msApp._dataSourceReferences["default.cds"].dataSources = null;
+            errors.ThrowOnErrors();
+
+            using var sourcesTempDir = new TempDir();
+            var sourcesTempDirPath = sourcesTempDir.Dir;
+            errors = msApp.SaveToSources(sourcesTempDirPath, pathToMsApp);
+            (msApp, errors) = CanvasDocument.LoadFromSources(sourcesTempDirPath);
+            using var msAppTemp = new TempFile();
+            using var sources2 = new TempDir();
+            var errors2 = new ErrorContainer();
+            MsAppSerializer.SaveAsMsApp(msApp, msAppTemp.FullPath, errors2);
+            errors = msApp.SaveToSources(sources2.Dir, msAppTemp.FullPath);
+            errors.ThrowOnErrors();
+            var entropy = ReadEntropy(sources2.Dir);
```

Comment:
Need to fix the file not found error here
<img width="1000" alt="image" src="https://github.com/microsoft/PowerApps-Language-Tooling/assets/98551644/49c538a3-dc15-4a0d-be1a-452874340c47">